crates/ruff_server/src/edit/notebook.rs

```diff
@@ -0,0 +1,189 @@
+use std::{collections::HashMap, hash::BuildHasherDefault};
+
+use anyhow::Ok;
+use lsp_types::{NotebookCellKind, Url};
+use rustc_hash::FxHashMap;
+
+use crate::{PositionEncoding, TextDocument};
+
+use super::DocumentVersion;
+
+/// The state of a notebook document in the server. Contains an array of cells whose
+/// contents are internally represented by [`TextDocument`]s.
+#[derive(Clone, Debug)]
+pub(crate) struct NotebookDocument {
+    cells: Vec<NotebookCell>,
+    metadata: ruff_notebook::RawNotebookMetadata,
+    version: DocumentVersion,
+    // Used to quickly find the index of a cell for a given URL.
+    cell_index: FxHashMap<lsp_types::Url, usize>,
+}
```

Comment:
nit: we could utilize `ruff_index::newtype_index` for `cells` and `cell_index`. This way the indexing would be in a type-safe manner. For example,

https://github.com/astral-sh/ruff/blob/4ddd930c60cdda26b495bff0e4a22c72b1c682ab/crates/ruff_python_semantic/src/binding.rs#L330-L342

In this case, it would be something like:
```rs
#[newtype_index]
pub struct CellId;

pub struct NotebookCells(IndexVec<CellId, NotebookCell>);
```

This way we use `CellId` instead of a random `usize` to index into the `NotebookCell` vector.

Reply:
I'm not sure if this approach works in this specific case, because `IndexVec` doesn't support insertions (which we need for notebook cell array updates).

Reply:
By 'insertions', I'm specifically referring to insertions at an index, like the `Vec::insert` method (which shifts up indexes after it).

Reply:
I see, thanks for clarifying. I leave this up to you but I think we should at least use a newtype wrapper (`pub type CellId = usize`) instead of raw `usize` as I've seen it used in a lot of places.

Reply:
Sure, we can do that!